src/Monarch/VirtualDom/VirtualDomTree.ts

```diff
@@ -315,6 +317,10 @@ function unsafe_diffElementNS<a, b>(x: VirtualDomTree.ElementNS<a>, y: VirtualDo
     const xChildrenLength = x.children?.length || 0
     const yChildrenLength = y.children?.length || 0
 
+    if (yChildrenLength > 0 && xChildrenLength > 0 && y.children![0].tag === VirtualDomTree.ElementNS && y.children![0].key) {
```

Comment:
I suggest to add a new sum type to the `VirtualDomTree` type named `KeyedElementNS`. We are able to determine at compile time that the children of an element in the virtual DOM have `key` or not. So as I mentioned in the https://github.com/the-dr-lazy/purescript-monarch/issues/18#issuecomment-742396828 we can make the `node` function ad-hoc polymorphic. If children has `key` fact we make `KeyedElementNS` otherwise we make `ElementNS`.

```purs
ul_ [ li {} [ ... ] ] :: VirtualDomTree Nothing slots message
-- {
--   tag: ElementNS
--   ...
-- }

ul_ [ li { key: user.id } [ ... ] ] :: VirtualDomTree Nothing slots message
-- {
--   tag: KeyedElementNS
--   ...
-- }


```